### PR TITLE
[AutoDiff] Part 3: Revamp DifferentiationTask

### PR DESCRIPTION
**Pull request by rxwei**
(Split from of #17901)

* Completely revamp `DifferentiationTask`. The new `DifferentiationTask` manages a `[reverse_differentiable]` attribute, stores the invoker of this task, and stores tasks associated with the current one.

* Make gradient function synthesis delete dead instructions such as the `function_ref` of the original function.

* Make `SILReverseAutoDiffIndices` and `SILReverseAutoDiffConfig` parameters pass by reference.

* Use `SILModule::lookUpWitnessTable` instead of holding a list of serialized witness tables locally.

* Move some local helper functions to the top of the file to eliminate declarations.

* Add underpinnings for differentiability diagnostics (backed by `DifferentiationInvoker`).

**Note:** The next patch, primal synthesis, will make unused function warnings go away.

### CODE REVIEW

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp [hunk 1]
```diff
@@ -22,162 +22,500 @@
...
-static NominalTypeDecl *getStdlibTypeDecl(StringRef, ASTContext &);
-static std::string mangleADIndices(SILReverseAutoDiffIndices);
-static std::string mangleADConfig(const SILReverseAutoDiffConfiguration &);
-static SILFunction *lookupOrLinkFunction(StringRef name, SILModule &);
-static FuncDecl *lookupAssociativeOperatorDeclInProtocol(DeclName operatorName,
-                                                         ProtocolDecl *);
-static void collectAllFormalResultsInTypeOrder(SILFunction &,
-                                               SmallVectorImpl<SILValue> &);
+/// Prints an "[AD] " prefix to `llvm::dbgs()` and returns the debug stream.
+/// This is being used to print short debug messages with the AD pass.
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: Typo: `with the AD pass` -> `within the AD pass`

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp [hunk 2]
```diff
@@ -349,46 +666,197 @@ class ADContext {
...
+  switch (invoker.getKind()) {
+  // For a gradient instruction that is not associated with any source
+  // location, we emit a diagnostic without source location.
+  case DifferentiationInvoker::Kind::GradientInst:
+    diagnose(srcLoc, diag::autodiff_function_not_differentiable);
+    return;
+
+  // For indirect differentiation, emit a "not differentiable" note on the
+  // expression first. Then emit an error at the source invoker of
+  // differentiation, and a "when differentiating this"  note at each indirect
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: extra space before `note`:
```
"when differentiating this"  note
```

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp [hunk 3]
```diff
@@ -349,46 +666,197 @@ class ADContext {
...
+    auto *outerTask = invoker.getIndirectDifferentiation().second;
+    while (outerTask->getInvoker().getKind() ==
+             DifferentiationInvoker::Kind::IndirectDifferentiation) {
+      std::tie(inst, outerTask) =
+        outerTask->getInvoker().getIndirectDifferentiation();
+      auto applyLoc = inst->getLoc().getSourceLoc();
+      if (applyLoc.isValid())
+        diagnose(applyLoc, diag::autodiff_when_differentiating_function_call);
+    }
+    // Now we've reached a direct task, recursive once to emit an error.
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: Typo: `recursive once` -> `recurse once`

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp [hunk 4]
```diff
@@ -1332,146 +1353,183 @@ static void convertFromIntegerLiteral(intmax_t value,
...
                       /*isNonThrowing*/ false);
 }
 
 /// Create a seed value.
 ///
-/// NOTE: This will reduced to only support scalar AD when vector AD supports
-/// optional seeds, because a vector of 1 doesn't make sense in vector AD.
+/// NOTE: This will be reduced to only support scalar AD when vector AD supports
+/// optional seeds, because a vector of 1s as seed doesn't make mathematical
+/// sense in vector AD.
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: Good to know, thanks for the comment.

**Review by dan-zheng — APPROVED** [review]

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
odd. it builds locally on my machine

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Pull request merged**